### PR TITLE
fix: Re-Implement `RelationalConsistency.PreventDeletion` validation

### PR DESCRIPTION
**Pull request by sveneberth**
You can set `consistency` on a `RelationalBone` to `RelationalConsistency.PreventDeletion` to disallow the deletion of the referenced Skeleton (if there is a relation from this bone to it).

This works fine as long a Skeleton is referenced by only few other Skeleton. But if thousands of Skeletons has a reference to the same Skeleton (10000:1 relation) the `viur_incomming_relational_locks` list has to hold 10000 keys, which exceeds the size limit of the entity. Furthermore the transaction limit would be exceeded during writing on 10000 entites the `"{bonename}_outgoingRelationalLocks"` property too.
We cannot carry a list, neither in the one way nor in the other. (Yes, okay, theoretically you could write _outgoingRelationalLocks outside the transaction, maybe in a task. But even here it can be dangerous if a skeleton contains many (multiple) `RelationalBone`. Even if the number of outgoing ones will probably be smaller than the number of incoming ones.  But this is still an insane overhead).

That's why I thought about what I could use to solve this. And in my opinion, we can simply use what we already have: The `viur-relations` entities. The `consistency` is already stored in this. 
So I have removed everything that relates to the maintenance of these lock lists, ensured that these lists will no longer be present in the datastore entites in future and replaced the "can delete" check with a query on the `viur-relations`.
In my tests this works very well and we should even save some runtime and memory.

### CODE REVIEW

File: src/viur/core/bones/relational.py [hunk 1]
```diff
@@ -482,29 +477,8 @@ def serialize(self, skel: 'SkeletonInstance', name: str, parentIndexed: bool) ->
...
-                    referencedObj["viur_incomming_relational_locks"].append(skel["key"])
-                    db.Put(referencedObj)
-
-        for oldLock in oldRelationalLocks - newRelationalLocks:
-            # Remove Lock
-            if referencedObj := db.Get(oldLock):
-                if skel["key"] in referencedObj["viur_incomming_relational_locks"]:
-                    referencedObj["viur_incomming_relational_locks"].remove(skel["key"])
-                    db.Put(referencedObj)
+        # Delete legacy property
```
Comment (phorward) [inline src/viur/core/bones/relational.py:480]:
```suggestion
        # FIXME VIUR4: Delete legacy property
```

File: src/viur/core/bones/relational.py [hunk 2]
```diff
@@ -620,25 +566,25 @@ def postSavedHandler(self, skel, boneName, key):
...
-        dbVals.filter("src.__key__ =", key)
-        db.Delete([x for x in dbVals.run()])
-
-    def isInvalid(self, key):
+        query = db.Query("viur-relations")
+        query.filter("viur_src_kind =", skel.kindName)
+        query.filter("viur_dest_kind =", self.kind)
+        query.filter("viur_src_property =", boneName)
+        query.filter("src.__key__ =", key)
+        db.Delete([x for x in query.run()])
```
Comment (phorward) [inline src/viur/core/bones/relational.py:585]:
```suggestion
        db.Delete([entity for entity in query.run()])
```

File: src/viur/core/bones/relational.py [hunk 3]
```diff
@@ -923,8 +869,10 @@ def buildDBSort(
         origFilter = dbFilter.queries
         if origFilter is None or not "orderby" in rawFilter:  # This query is unsatisfiable or not sorted
             return dbFilter
-        if "orderby" in rawFilter and isinstance(rawFilter["orderby"], str) and rawFilter["orderby"].startswith(
-                f"{name}."):
+        if (
+            "orderby" in rawFilter and isinstance(rawFilter["orderby"], str)
+            and rawFilter["orderby"].startswith(f"{name}.")
+        ):
```
Comment (phorward) [inline src/viur/core/bones/relational.py:875]: Can you please keep the ugly code, as it is refactored by #1077 already for viur-core 3.7 and results in a merge conflict.

```suggestion
        if "orderby" in rawFilter and isinstance(rawFilter["orderby"], str) and rawFilter["orderby"].startswith(
                f"{name}."):
```

File: src/viur/core/bones/relational.py [hunk 4]
```diff
@@ -1217,27 +1163,33 @@ def setBoneValue(
...
+            ):
                 raise ValueError(f"You must supply a tuple of (Database-Key, relSkel) to {boneName}")
             realValue = value
         elif self.multiple and not self.using:
-            if not (isinstance(value, str) or isinstance(value, db.Key)) and not (isinstance(value, list)) \
-                    and all([isinstance(x, str) or isinstance(x, db.Key) for x in value]):
+            if (
+                not (isinstance(value, str) or isinstance(value, db.Key))
+                and not (isinstance(value, list))
+                and all(isinstance(x, (str, db.Key)) for x in value)
```
Comment (phorward) [inline src/viur/core/bones/relational.py:1177]:
```suggestion
                and all(isinstance(k, (str, db.Key)) for k in value)
```

File: src/viur/core/skeleton.py [hunk 5]
```diff
@@ -1048,6 +1049,9 @@ def txnUpdate(write_skel):
                         else:
                             logging.critical("Detected Database corruption! Could not delete stale lock-object!")
 
+            # Delete legacy property
```
Comment (phorward) [inline src/viur/core/skeleton.py:1052]:
```suggestion
            # FIXME VIUR4: Delete legacy property
```

File: src/viur/core/skeleton.py [hunk 6]
```diff
@@ -1404,14 +1416,17 @@ def __init__(self, baseSkel=None):
...
+        .filter("dest.__key__ =", removedKey)
+        .filter("viur_relational_consistency >", RelationalConsistency.PreventDeletion.value)
+    )
     updateListQuery = updateListQuery.setCursor(cursor)
     updateList = updateListQuery.run(limit=5)
     for entry in updateList:
         skel = skeletonByKind(entry["viur_src_kind"])()
         assert skel.fromDB(entry["src"].key)
-        if entry["viur_relational_consistency"] == 3:  # Set Null
+        if entry["viur_relational_consistency"] == RelationalConsistency.SetNull.value:
```
Comment (phorward) [inline src/viur/core/skeleton.py:1429]: lol! :+1:

**Review by phorward — CHANGES_REQUESTED** [review]
Hello @sveneberth, this is a very nice solution!
Thanks for kicking out this ugly and misspelled `viur_incomming_relational_locks`...

Can you please take a look to my comments, it's only minor stuff.

File: src/viur/core/bones/relational.py (hunk 3 above)
Comment (sveneberth) [reply src/viur/core/bones/relational.py:875]: well okay. The linter always disagrees with PyCharms, so I'll probably have to mediate manually again.

Comment (phorward) [reply src/viur/core/bones/relational.py:875]: well okay, then leave it as is, I'll manually resolve the conflict.

**Review by phorward — DISMISSED** [review]
I'm fine with this, although it are very huge change for a "bugfix release" 3.6.14 ;)

**Review by phorward — APPROVED** [review]

**Comment by sveneberth** [comment]
> I'm fine with this, although it are very huge change for a "bugfix release" 3.6.14 ;)

#1240 and #1228 were also huge :) SemVer hasn't been here at viur-core for a long time anyway… :/

**Pull request merged**